File: backend/supabase/chatService.js
import { supabase } from './supabase.js';

/**
 * Service for handling chat operations with Supabase
 */
export class ChatService {
  /**
   * Create or retrieve an existing chat session
   * @param {string} userId - User ID
   * @param {string} platform - Platform identifier (webchat, whatsapp, etc.)
   * @param {string} externalSessionId - External platform session ID (optional)
   * @returns {Promise<Object>} - Session data
   */
  async createOrFindSession(userId, platform, externalSessionId = null) {
    try {
      const { data, error } = await supabase.rpc('find_or_create_chat_session', {
        p_user_id: userId,
        p_platform: platform,
        p_external_session_id: externalSessionId
      });

      if (error) throw error;
      return { sessionId: data };
    } catch (error) {
      console.error('Error creating/finding chat session:', error);
      throw error;
    }
  }

  /**
   * End a chat session
   * @param {string} sessionId - Session ID
   * @returns {Promise<Object>} - Success status
   */
  async endSession(sessionId) {
    try {
      const { data, error } = await supabase
        .from('chat_sessions')
        .update({
          is_active: false,
          ended_at: new Date()
        })
        .eq('id', sessionId);

      if (error) throw error;
      return { success: true };
    } catch (error) {
      console.error('Error ending chat session:', error);
      throw error;
    }
  }

  /**
   * Record a chat message
   * @param {string} sessionId - Session ID
   * @param {string} senderType - 'user' or 'assistant'
   * @param {string} content - Message content
   * @param {Object} metadata - Additional metadata (optional)
   * @returns {Promise<Object>} - Message data
   */
  async recordMessage(sessionId, senderType, content, metadata = null) {
    try {
      const { data, error } = await supabase.rpc('record_chat_message', {
        p_session_id: sessionId,
        p_sender_type: senderType,
        p_content: content,
        p_metadata: metadata
      });

      if (error) throw error;
      return { messageId: data };
    } catch (error) {
      console.error('Error recording chat message:', error);
      throw error;
    }
  }

  /**
   * Get chat history for a session
   * @param {string} sessionId - Session ID
   * @returns {Promise<Array>} - Chat messages
   */
  async getChatHistory(sessionId) {
    try {
      const { data, error } = await supabase
        .from('chat_messages')
        .select('*')
        .eq('session_id', sessionId)
        .order('sent_at', { ascending: true });

      if (error) throw error;
      return data;
    } catch (error) {
      console.error('Error fetching chat history:', error);
      throw error;
    }
  }

  /**
   * Record user intent
   * @param {string} userId - User ID
   * @param {string} sessionId - Session ID
   * @param {string} intentType - Intent type (e.g., 'compare_fees', 'project_balance')
   * @param {Object} intentData - Intent-specific data
   * @returns {Promise<Object>} - Intent record
   */
  async recordIntent(userId, sessionId, intentType, intentData) {
    try {
      const { data, error } = await supabase.rpc('record_user_intent', {
        p_user_id: userId,
        p_session_id: sessionId,
        p_intent_type: intentType,
        p_intent_data: intentData
      });

      if (error) throw error;
      return { intentId: data };
    } catch (error) {
      console.error('Error recording user intent:', error);
      throw error;
    }
  }

  /**
   * Get recent sessions for a user
   * @param {string} userId - User ID
   * @param {number} limit - Maximum number of sessions to retrieve
   * @returns {Promise<Array>} - Session data
   */
  async getUserSessions(userId, limit = 10) {
    try {
      const { data, error } = await supabase
        .from('chat_sessions')
        .select('*, chat_messages(count)')
        .eq('user_id', userId)
        .order('started_at', { ascending: false })
        .limit(limit);

      if (error) throw error;
      return data;
    } catch (error) {
      console.error('Error fetching user sessions:', error);
      throw error;
    }
  }
}

File: backend/supabase/chatService.py
import uuid
from .supabase import supabase

class ChatService:
    """Service for handling chat operations with Supabase"""
    
    async def createOrFindSession(self, user_id, platform, external_session_id=None):
        """
        Create or retrieve an existing chat session
        """
        try:
            data = {
                "p_user_id": user_id,
                "p_platform": platform,
                "p_external_session_id": external_session_id
            }
            
            # Call the Supabase function
            result = await supabase.query(
                "/rest/v1/rpc/find_or_create_chat_session",
                method="POST",
                data=data
            )
            
            return {"sessionId": result}
        except Exception as e:
            print(f"Error creating/finding chat session: {e}")
            # Fallback: generate a local session ID
            return {"sessionId": str(uuid.uuid4())}
    
    async def endSession(self, session_id):
        """End a chat session"""
        try:
            data = {
                "is_active": False,
                "ended_at": "now()"
            }
            
            await supabase.query(
                f"/rest/v1/chat_sessions?id=eq.{session_id}",
                method="PATCH",
                data=data
            )
            
            return {"success": True}
        except Exception as e:
            print(f"Error ending chat session: {e}")
            return {"success": False, "error": str(e)}
    
    async def recordMessage(self, session_id, sender_type, content, metadata=None):
        """Record a chat message"""
        try:
            data = {
                "p_session_id": session_id,
                "p_sender_type": sender_type,
                "p_content": content,
                "p_metadata": metadata
            }
            
            # Call the Supabase function
            result = await supabase.query(
                "/rest/v1/rpc/record_chat_message",
                method="POST",
                data=data
            )
            
            return {"messageId": result}
        except Exception as e:
            print(f"Error recording chat message: {e}")
            # Fallback: return a generated ID
            return {"messageId": str(uuid.uuid4())}
    
    async def getChatHistory(self, session_id):
        """Get chat history for a session"""
        try:
            result = await supabase.query(
                f"/rest/v1/chat_messages?session_id=eq.{session_id}",
                method="GET"
            )
            
            return result
        except Exception as e:
            print(f"Error fetching chat history: {e}")
            return []
    
    async def recordIntent(self, user_id, session_id, intent_type, intent_data):
        """Record user intent"""
        try:
            data = {
                "p_user_id": user_id,
                "p_session_id": session_id,
                "p_intent_type": intent_type,
                "p_intent_data": intent_data
            }
            
            # Call the Supabase function
            result = await supabase.query(
                "/rest/v1/rpc/record_user_intent",
                method="POST",
                data=data
            )
            
            return {"intentId": result}
        except Exception as e:
            print(f"Error recording user intent: {e}")
            return {"intentId": str(uuid.uuid4())}